lib/HttpClient.ts

```diff
@@ -186,48 +200,68 @@ export class HttpClient implements ifm.IHttpClient {
-        if (response && response.message && response.message.statusCode === HttpCodes.Unauthorized) {
-            let authenticationHandler: ifm.IRequestHandler;
+        // Only perform retries on reads since writes may not be idempotent.
+        let maxTries = (this._allowRetries && RetryableHttpVerbs.indexOf(verb) != -1) ? this._maxRetries + 1 : 1;
+        let numTries = 0;
+
+        let response: HttpClientResponse;
+        while (numTries < maxTries) {
+            response = await this.requestRaw(info, data);
+
+            // Check if it's an authentication challenge
+            if (response && response.message && response.message.statusCode === HttpCodes.Unauthorized) {
+                let authenticationHandler: ifm.IRequestHandler;
+
+                for (let i = 0; i < this.handlers.length; i++) {
+                    if (this.handlers[i].canHandleAuthentication(response)) {
+                        authenticationHandler = this.handlers[i];
+                        break;
+                    }
+                }
+
+                if (authenticationHandler) {
+                    return authenticationHandler.handleAuthentication(this, info, data);
+                }  
+                else {
+                    // We have received an unauthorized response but have no handlers to handle it.
+                    // Let the response return to the caller.
+                    return response;
+                }
+            }
+
+            let redirectsRemaining: number = this._maxRedirects;
+            while (HttpRedirectCodes.indexOf(response.message.statusCode) != -1
+                && this._allowRedirects
+                && redirectsRemaining > 0) {
 
-            for (let i = 0; i < this.handlers.length; i++) {
-                if (this.handlers[i].canHandleAuthentication(response)) {
-                    authenticationHandler = this.handlers[i];
+                const redirectUrl: any = response.message.headers["location"];
+                if (!redirectUrl) {
+                    // if there's no location to redirect to, we won't
                     break;
                 }
+
+                // we need to finish reading the response before reassigning response
+                // which will leak the open socket.
+                await response.readBody();
+
+                // let's make the request with the new redirectUrl
+                info = this._prepareRequest(verb, redirectUrl, headers);
+                response = await this.requestRaw(info, data);
+                redirectsRemaining--;
             }
 
-            if (authenticationHandler) {
-                return authenticationHandler.handleAuthentication(this, info, data);
-            }  
-            else {
-                // We have received an unauthorized response but have no handlers to handle it.
-                // Let the response return to the caller.
+            if (HttpResponseRetryCodes.indexOf(response.message.statusCode) == -1) {
```

Comment:
&& check RetryableHttpVerbs? Only retry if in there. Let's add tests to make sure we don't retry on not retryable verbs, regardless of response.

Reply:
We don't need to check `RetryableHttpVerbs` because of this expression above: `let maxTries = (this._allowRetries && RetryableHttpVerbs.indexOf(verb) != -1) ? this._maxRetries + 1 : 1;`

If its not in `RetryableHttpVerbs`, `maxTries` will equal 1 and it will fall out of the loop.

Reply:
Added more testing